crates/storage/src/state_update.rs

```diff
@@ -0,0 +1,194 @@
+
+        let nonce = context
+            .intern_felt(nonce.0)
+            .context("Interning nonce value")?;
+
+        context
+            .insert_nonce(block_number, contract, nonce)
+            .context("Inserting nonce update")?;
+    }
+
+    // // Insert storage updates
+    for StorageDiff {
+        address,
+        key,
+        value,
+    } in &state_diff.storage_diffs
+    {
+        // Contract address should already have be present by now so no need to intern.
+        let contract = context
+            .get_felt_id(*address.get())
+            .context("Reading storage update's contract class ID")?;
+
+        let key = context
+            .intern_felt(*key.get())
+            .context("Interning storage key")?;
+
+        let value = context
+            .intern_felt(value.0)
+            .context("Interning storage value")?;
+
+        context
+            .insert_storage_update(block_number, contract, key, value)
+            .context("Inserting storage update")?;
+    }
+
+    Ok(())
+}
+
+mod sql {
+    use anyhow::Context;
+    use pathfinder_common::StarknetBlockNumber;
+    use rusqlite::{Statement, Transaction};
+    use stark_hash::Felt;
+
+    pub struct SqlContext<'tx> {
+        query_felt: Statement<'tx>,
+        insert_felt: Statement<'tx>,
+        insert_nonce: Statement<'tx>,
+        insert_storage: Statement<'tx>,
+        insert_contract: Statement<'tx>,
+    }
+
+    impl<'tx> SqlContext<'tx> {
+        pub fn new(tx: &'tx Transaction<'tx>) -> anyhow::Result<Self> {
+            let insert_felt = tx
+                .prepare("INSERT OR IGNORE INTO felts (id, felt) VALUES (NULL, ?)")
+                .context("Preparing felt insert statement")?;
+
+            let query_felt = tx
+                .prepare("SELECT id FROM felts WHERE felt = ?")
+                .context("Preparing felt query statement")?;
+
+            let insert_nonce = tx
+                .prepare(
+                    "INSERT INTO nonce_updates (block_number, contract_address, nonce) VALUES (?, ?, ?)",
+                )
+                .context("Preparing nonce insert statement")?;
+
+            let insert_storage = tx
+                .prepare("INSERT INTO storage_updates (block_number, contract_address, storage_address, storage_value) VALUES (?, ?, ?, ?)")
+                .context("Preparing nonce insert statement")?;
+
+            let insert_contract = tx
+                .prepare("INSERT INTO contract_updates (block_number, contract_address, class_hash) VALUES (?, ?, ?)")
+                .context("Preparing contract insert statement")?;
+
+            Ok(Self {
+                insert_felt,
+                query_felt,
+                insert_nonce,
+                insert_storage,
+                insert_contract,
+            })
+        }
+
+        pub fn intern_felt(&mut self, felt: Felt) -> anyhow::Result<u64> {
+            self.insert_felt.execute([felt.as_be_bytes()])?;
+            self.get_felt_id(felt)
```

Comment:
I was hoping these could be combined using `RETURNING id`, but unfortunately not possible with `OR IGNORE` as it doesn't actually alter the record. One could do it with `OR REPLACE` but I think it does not matter much.

Reply:
No longer applicable as we are not going to use interning.